### ISSUE TITLE
NanoSVG Style Sheets

### ISSUE THREAD

**Issue by andrewbest** [VCVRack/Rack#577]
There is a version of NanoSVG that addresses an issue with style sheets. The problem is that files from Adobe Illustrator (or other inline Style Sheets) will not render. The issue is tracked here: https://github.com/memononen/nanosvg/issues/55. I have tested the file provided by the last poster: https://github.com/olivierchatry/svg-auto-viewbox/blob/master/Externals/nanosvg/nanosvg.h. It works for me and renders my Adobe Illustrator test file. I request that this file replace the current version of nanosvg.h. This would expand the supported SVG editors making it easier for developers.

Note: You have to add #include <float.h> to the poster's file.

**Comment by AndrewBelt** [comment]
What if memononen makes a change to NanoSVG that is not updated in olivierchatry's version? Who will maintain olivierchatry's file to always be updated?

**Comment by andrewbest** [comment]
Maybe he would accept a pull request. Currently, Adobe Illustrator is not supported, which seems pretty bad.

**Comment by AndrewBelt** [comment]
I don't value having this feature, but I'd be happy to update the nanosvg submodule if olivierchatry's fix makes it to nanosvg master.

**Comment by andrewbest** [comment]
What SVG editors do you support for developing VCV Rack modules?

**Comment by AndrewBelt** [comment]
I support Inkscape, and anything else that generates sane SVG should work also. There is a way to generate better SVG with Illustrator, which you can find by searching the Facebook User Group, but I don't own or use Illustrator so I can't help you on that.

**Comment by squinkylabs** [comment]
+1 for this. And it's not the that Illustrator makes "insane" svg, although maybe it does. The issue is that VCV can't read SGV styles unless they are inline. So this fragment doesn't import:
```  <defs>
    <style>
      .cls-1 {
        fill: #19ea82;
      }

      .cls-5 {
        stroke: none;
      }

      .cls-6 {
        fill: none;
      }
    </style>
  </defs>
  <g id="Rectangle_1" data-name="Rectangle 1" class="cls-1">
    <rect class="cls-5" width="25" height="23"/>
    <rect class="cls-6" x="0.5" y="0.5" width="24" height="22"/>
  </g>```
I think that's pretty sane for SVG. Inkscape will read it in just fine, and every browser renders it OK. Sounds like nanosvg just has a bug in it. A quick look at the source tells me they have code to parse styles..

**Comment by andrewbest** [comment]
You can use these settings in Illustrator.
- Export As or Artboards Export
- Styling: Inline Style
- Font: Convert to Outlines
- Images: Embed
- Object IDs: Layer Names
- Minify unchecked
- Responsive unchecked

**Comment by AndrewBelt** [comment]
>Sounds like nanosvg just has a bug in it

This "bug" is not having support for the entire CSS standard, so I think we can cut the author some slack on that one.

**Comment by andrewbest** [comment]
Asset Export in the Artboards window is a quick way to export svg files with saved settings. It has saved me a lot of time.

**Comment by squinkylabs** [comment]
yes, @andrewbest I use your recommendations and it works great. In fact I can read in "bad" SVG and then export with your settings and everything is good to go.

@AndrewBelt as I said, it appears that nanosvg actually does try to parse style tags. So I'm hoping a fix will come. Even if you hate Illustrator, it's hard to deny it's probably the most popular way to  create SVG. I doubt that nano would intentionally not support a filled rectangle using the style tag, but I guess anything is possible. I'll report back when I know more.

**Comment by AndrewBelt** [comment]
@squinkylabs Style tags are a small subset of CSS, and NanoSVG can only parse a small subset of style tags, something like 20 attributes if written in a sane way (no comments, corner cases of CSS, etc). Please send $70,000 to my PayPal account so I can forward it to memononen so he can implement this feature.

**Comment by squinkylabs** [comment]
@AndrewBelt : Hah - point well taken! I also realized that in the spirit of being lean and mean (small and fast) maybe one wouldn't even want to support the "defs" section of SVG. Perhaps I can get someone to send my paypal 70k to write a conversion tool to make it easier to use SVG found in the wild with nano.
